File: web/dartScripts/selectSchoolFromList.dart
import 'dart:html';

void main() {
  var targetDiv = querySelector('#schoolList');
  var inputField = querySelector('#urlSelectField');

  inputField.onInput.listen(handleNewLetterInUrlSelectField);
}

bool requestIsAlreadyRunning = false;
void handleNewLetterInUrlSelectField(Event ev) {
  var newValue = ev.matchingTarget.nodeValue.trim();

  if (newValue.startsWith(RegExp('http(s)?://'))) {
    return;
  }

  // Ask server for school list
  if (!requestIsAlreadyRunning) {
    requestIsAlreadyRunning = true;
    // Yeah, we could see two requests at once because async, but who cares
    var request = HttpRequest.request('/schoolList.csv?query=' + Uri.encodeQueryComponent(newValue))
        .then((req) => updateSchoolList(req));
  }
}

void updateSchoolList(HttpRequest req) {
  var targetDiv = querySelector('#schoolList');

  var lines = req.responseText.split('\n').map((l) => l.split(':'));
  for (var line in lines) {
    var name = line[0];
    var url = line[1];
    var newElement = Element.div();
    newElement.appendText(name);
    newElement.className = "schoolListRow";
    newElement.setAttribute('url', url);
  }
}

void handleClickOnSchoolInSchoolList(Event ev) {}

File: web/js/selectSchoolFromList.js
var inputField = document.getElementById('urlSelectField');
var targetDiv = document.getElementById('schoolList');

inputField.addEventListener('input', handleNewLetterInUrlSelectField);

var requestIsAlreadyRunning = false;
function handleNewLetterInUrlSelectField(e) {
    var inputValue = inputField.value.trim();

    if (inputValue == '' || inputValue.startsWith('http://') || inputValue.startsWith('https://'))
        return;

    if (!requestIsAlreadyRunning) {
        requestIsAlreadyRunning = true;
        var url = '/schoolList.csv?query=' + encodeURIComponent(inputValue);
        var request = new XMLHttpRequest();
        request.open('GET', url, true);
        request.onload = function () {
            if (this.status >= 200 && this.status < 400) {
                var resp = this.responseText;
                var lines = resp.split('\n');
                targetDiv.innerHTML = '';
                lines.forEach(line => {
                    var name = line.split('``````')[0];
                    var url = line.split('``````')[1];
                    targetDiv.innerHTML += '<div class="schoolListRow" url="' + url + '" onclick="handleItemClick(\'' + url + '\')">' + name + '</div>'
                });
            } else {
                console.log('Error fetching url ' + url + ' . Status code: ' + this.status);
            }
            requestIsAlreadyRunning = false;
        }
        request.send();
    }
}

function handleItemClick(url) {
    inputField.value = url;
    targetDiv.innerHTML = '';
}